src/test/java/com/cloudbees/jenkins/support/configfiles/SecretHandlerTest.java

```diff
@@ -88,4 +91,28 @@ public void shouldPutAPlaceHolderInsteadOfSecret() throws Exception {
         FileUtils.writeStringToFile(file, xml);
         String patchedXml = SecretHandler.findSecrets(file);
         assertEquals(expectedXml, patchedXml);
-    }}
+    }
+
+    @Test
+    public void shouldNotResolveExternalEntities() throws Exception {
+        final String xxeXml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n" +
+                "<!DOCTYPE test [ \n" +
+                "    <!ENTITY xxeattack SYSTEM \"file:///\"> \n" +
+                "]>\n" +
+                "<xxx>&xxeattack;</xxx>";
+        File file = File.createTempFile("test", ".xml");
+        FileUtils.writeStringToFile(file, xxeXml);
+        boolean fallbackEnabled = SecretHandler.ENABLE_FALLBACK;
+        try {
+            // Disable the fallback so TransformerExceptions are always thrown.
+            SecretHandler.ENABLE_FALLBACK = false;
+            String redactedXxeXml = SecretHandler.findSecrets(file);
+            // If there is no exception than the XML library should have removed the entity without processing it.
+            assertThat(redactedXxeXml, containsString("<xxx/>"));
+        } catch (TransformerException e) {
+            assertThat(e.getMessage(), containsString("Refusing to resolve entity"));
```

Comment:
The expected execution flow is not easily understandable and so the comment becomes mandatory to be read.

WDYT about this approach that explicitly show the flow ? (if/else against try/catch)

```
String redactedXxeXml = null;
TransformerException exceptionIfNotProcessable = null;
try {
    // Disable the fallback so TransformerExceptions are always thrown.
    SecretHandler.ENABLE_FALLBACK = false;
    redactedXxeXml = SecretHandler.findSecrets(file);
} catch (TransformerException e) {
    exceptionIfNotProcessable = e;
} finally {
    SecretHandler.ENABLE_FALLBACK = fallbackEnabled;
}

if(redactedXxeXml == null){
    // the parser supports the secure features
    assertNotNull(exceptionIfNotProcessable);
    assertThat(exceptionIfNotProcessable.getMessage(), containsString("Refusing to resolve entity"));
}else{
    // If there is no exception than the XML library should have removed the entity without processing it.
    assertNull(exceptionIfNotProcessable);
    assertThat(redactedXxeXml, containsString("<xxx/>"));
}
```

With that approach we tell the reader that both path are possible. With the try/catch approach, you need to read the comment to understand that actually it's not the last assert that can throw the exception but the findSecrets.

It's just a matter of perspective, do not feel forced to follow my subjective PoV.

Reply:
I agree that it is confusing to keep track of what is going on here. I will refactor it similarly to your recommendation to make it easier to understand the various cases.